### PR TITLE
[SPARK-26894][SQL] Handle Alias as well in AggregateEstimation to propagate child stats

### PR DESCRIPTION
**Pull request by venkata91**
## What changes were proposed in this pull request?

Currently aliases are not handled in the Aggregate Estimation due to which stats are not getting propagated. This causes CBO join-reordering to not give optimal join plans. ProjectEstimation is already taking care of aliases, we need same logic for AggregateEstimation as well to properly propagate stats when CBO is enabled.

## How was this patch tested?

This patch is manually tested using the query Q83 of TPCDS benchmark (scale 1000)

### CODE REVIEW

**Comment by venkata91** [comment]
cc @cloud-fan @wzhfy Please take a look at this change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: can we add a method in `EstimationUtils` in order to avoid the code redundancy?

**Review by mgaido91 — COMMENTED** [review]
The change seems reasonable to me, just left one comment. Also, please add a UT for this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala
Comment (venkata91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: Done. Also added unit test case.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala]: nit: please follow the usual pattern:
```
/**
 * comment...
 */
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala]: nit: indent (also the following line, please see the other methods)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: unneeded import

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/AggregateEstimationSuite.scala
Comment (mgaido91) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/AggregateEstimationSuite.scala]: since we are actually interested in checking only the `attributeStats`, what about `assert(testAgg.stats.attributeStats == expectedAttrStats)`? This way we can also skip creating the `Statistics` class above..

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/AggregateEstimationSuite.scala
Comment (mgaido91) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/AggregateEstimationSuite.scala]: super nit: `SPARK-26894: propagate child stats for aliases in Aggregate`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/AggregateEstimationSuite.scala
Comment (venkata91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/AggregateEstimationSuite.scala]: Makes sense. Done.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/AggregateEstimationSuite.scala
Comment (venkata91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/AggregateEstimationSuite.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala
Comment (venkata91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala
Comment (venkata91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala]: nit: `Returns the stats for aliases of child's attributes.`

**Review by mgaido91 — COMMENTED** [review]
just one minor comment, otherwise LGTM.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala
Comment (venkata91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala]: Done

**Comment by venkata91** [comment]
@mgaido91 can this be merged now?

**Comment by mgaido91** [comment]
@venkata91 I am fine with this change, I gave my LGTM, so I am ok merging this,, but I am not a committer, so the PR needs the review and approval of a committer who can merge it.

**Comment by venkata91** [comment]
@gatorsmile @cloud-fan Can one of you please take a look at this PR?

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by dilipbiswal** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala]: `ProjectEstimation` has the same logic and it can use this func?

Comment (venkata91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala]: Yeah, already changed that as part of the PR.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala]: oh, I see.

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
cc: @dongjoon-hyun

**Comment by venkata91** [comment]
@mgaido91 Can you please approve if everything looks good?

**Comment by dongjoon-hyun** [comment]
Thank you for cc-ing me, @maropu . I'll take a look.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (pending Jenkins), too. This PR works in the same way with unaliased attribute case.

@venkata91 . Could you share your Q83 TPCDS benchmark result in the PR description? Currently, it doesn't give any clear idea in both positive or negative way.
> This patch is manually tested using the query Q83 of TPCDS benchmark (scale 1000)

And, we usually use the following sentence.
> Pass the Jenkins with the newly added test case.

@maropu . I know that you already approved this and I'll leave this PR to you. Please merge this when you want after Jenkins passes.

**Comment by dongjoon-hyun** [comment]
BTW, thank you for your first contribution, @venkata91 !

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by venkata91** [comment]
> BTW, thank you for your first contribution, @venkata91 !

Thanks @dongjoon-hyun

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by maropu** [comment]
@dongjoon-hyun I couldn't assign @venkata91 in the jira and I'm not sure how to add a new account in contributor lists there... any info?

**Comment by dongjoon-hyun** [comment]
Sure, @maropu . For the first contributor, you need to him into `Apache Spark Contributor` group.

1. Visit Project Configuration Role Page
    - https://issues.apache.org/jira/plugins/servlet/project-config/SPARK/roles

2. Click `Add users to a role` button.

3. Fill the the pop-up box.
    - The JIRA user ID or email
    - Choose `Contributor`.

**Comment by maropu** [comment]
Thanks, @dongjoon-hyun! I could update.

**Pull request closed without merging**
